refsheet.cls

```diff
@@ -1,57 +1,212 @@
+%&tex
 \NeedsTeXFormat{LaTeX2e}
-\ProvidesClass{refsheet}[2013/02/25 Reference Sheet Class]
-
-
-\newcommand{\refsheet@columns}{3}
-
-% Would be good to have an option for how many columns are desired
-
-\DeclareOption*{%
-  \PassOptionsToClass{\CurrentOption}{article}%
+\ProvidesClass{refsheet}[2018/08/29 v1.0 Reference Sheet Class]
+
+%%% OPTION DEFINITIONS
+%
+%   Simple definitions for key-value pairs with kvoptions
+%   See J. Wright, C. Feuersänger, TUGBoat, Vol. 30 (2009), No. 1, p. 110-122
+%
+\RequirePackage{kvoptions}
+\SetupKeyvalOptions{
+  family=refsheet,
+  prefix=refsheet@
 }
+%
+%   Defining how many columns shall be used, 3 is the initial value,
+%   which can be set with rscols=<int> in the preamble.
+%   Disable this option for the body of the document.
+%
+\DeclareStringOption[3]{rscols}
+\AtBeginDocument{%
+  \DisableKeyvalOption[action=error,package=refsheet]{refsheet}{rscols}}
+%
+%   Define a preset margin for geometry
+%
+\DeclareStringOption[1cm]{margin}
+%
+%   Warn about options which will not be used/ passed on because of potential clashes
+%
+\DeclareVoidOption{portrait}{%
+  \PackageWarning{refsheet}{Incompatible with portrait mode}}
+\DeclareVoidOption{twocolumn}{%
+  \PackageError{refsheet}{Using 'multicol' for columns}}
+\DeclareVoidOption{titlepage}{%
+  \PackageWarning{refsheet}{Incompatible with a title page}}
+%
+%   Pass all other options to article
+%
+\DeclareDefaultOption{%
+  \PassOptionsToClass{\CurrentOptionKey}{article}}
+%
+%   DEV: Space for more options
+%
+
+%
+%   Process these options
+%   
+\ProcessKeyvalOptions{refsheet}
+%
+%%%
+
+%%% INHERIT STANDARD CLASS
+%
+%   Latest update of standard class article
+%   (It probably also works with earlier versions, but why sweat it.)
+\LoadClass{article}[2014/09/29]
+%
+%%%
+
+%%% ADDITIONAL REQUIRED TOOLS
+%
+%   Latest update of the nag package. 
+%   Harasses the user about deprecated usages.
+%
+\RequirePackage[orthodox,l2tabu,abort]{nag}[2011/11/25] 
+%
+%   Package etoolbox is necessary for additional hooks like \AtEndPreamble
+%
+\RequirePackage{etoolbox}
+%
+%   Page layout with latest version of geometry
+%
+\RequirePackage{geometry}[2018/04/16]
+\AtEndPreamble{%
+  \geometry{landscape,margin=\refsheet@margin}}
+%
+%   Produce columns with multicol
+%
+\RequirePackage{multicol}[2018/04/20]
+\AfterEndPreamble{%
```

Comment:
Hm, Why is this `\AfterEndPreamble`? My TeX days are far behind (and hopefully still ahead) of me, but I thought for sure there was an `\AtBeginDocument`. Wouldn't that make more sense? Content (including opening/closing environments) should be in the main document body.

Reply:
I thought so, too, but then I ran into a problem loading the `xcolor` package for the listings environment. Heiko explains that in [Undefined control sequence with Color Package in Custom LaTeX document class](https://tex.stackexchange.com/a/102998/33413), here quoted for convenience:

> The class uses `\AtBeginDocument` for typesetting. But this is not allowed, from LaTeX Companion, 2nd ed., p. 884:
>
>> Note, however, that code in the `\AtBeginDocument` hook is part of the preamble. Thus, restrictions limit what can be put there; in particular, no typesetting can be done.
>
> In this case, package color has not yet finished its setup, defining `\defaultcolor`. The typesetting code calls `\normalcolor` that assigns `\current@color` to the yet undefined `\defaultcolor`.

So this is necessary to avoid any changes lying ahead (and maybe for a future extension using colours).